hazelcast/src/main/java/com/hazelcast/map/impl/operation/EntryOperation.java

```diff
@@ -248,76 +251,157 @@ private void throwModificationInReadOnlyException() {
+        }
+    }
+
+    private void unlock(Data finalDataKey, String finalCaller, long finalThreadId, long finalCallId, Throwable cause) {
+        boolean unlocked = recordStore.unlock(finalDataKey, finalCaller, finalThreadId, finalCallId);
+        if (!unlocked) {
+            throw new IllegalStateException(
+                    String.format("Could not unlock by the caller=%s and threadId=%d", finalCaller, threadId), cause);
+        }
+    }
+
+    @SuppressWarnings({"unchecked", "checkstyle:methodlength"})
     private void updateAndUnlock(Data previousValue, Data newValue, EntryEventType modificationType, String caller,
                                  long threadId, final Object result, long now) {
         EntryOffloadableSetUnlockOperation updateOperation = new EntryOffloadableSetUnlockOperation(name, modificationType,
                 dataKey, previousValue, newValue, caller, threadId, now, entryProcessor.getBackupProcessor());
 
         updateOperation.setPartitionId(getPartitionId());
-        ops.invokeOnPartition(updateOperation).andThen(new ExecutionCallback<Object>() {
+        updateOperation.setReplicaIndex(0);
+        updateOperation.setNodeEngine(getNodeEngine());
+        updateOperation.setCallerUuid(getCallerUuid());
+        OperationAccessor.setCallerAddress(updateOperation, getCallerAddress());
+        @SuppressWarnings("checkstyle:anoninnerlength")
+        OperationResponseHandler setUnlockResponseHandler = new OperationResponseHandler() {
             @Override
-            public void onResponse(Object response) {
-                try {
-                    getOperationResponseHandler().sendResponse(EntryOperation.this, result);
-                } finally {
-                    ops.onCompletionAsyncOperation(EntryOperation.this);
+            public void sendResponse(Operation op, Object response) {
+                if (isRetryable(response) || isTimeout(response)) {
+                    retry(op);
+                } else {
+                    handleResponse(response);
                 }
             }
 
-            @Override
-            public void onFailure(Throwable t) {
-                try {
-                    // EntryOffloadableLockMismatchException is a marker send from the EntryOffloadableSetUnlockOperation
-                    // meaning that the whole invocation of the EntryOffloadableOperation should be retried
-                    if (t instanceof EntryOffloadableLockMismatchException) {
-                        t = new RetryableHazelcastException(t.getMessage(), t);
+            private void retry(final Operation op) {
+                setUnlockRetryCount++;
+                if (isRetryLimitReached()) {
+                    try {
+                        Object exceeded = new HazelcastException("Set & Unlock retry limit exceeded for key " + dataKey);
+                        getOperationResponseHandler().sendResponse(EntryOperation.this, exceeded);
```

Comment:
I was thinking about it thoroughly before, and the reasoning is as follows:
- I need to get to the partition thread the to unlock it, if I can't do it, how can I force unlock it?
I could try doing it, but I may not be on the partition thread, nor the partition may be available for me (because it got migrated away or sth similar).
- in theory, if there's no migrations and there is no hazelcast internal error this operation should eventually succeed, so maybe we should just retry indefinitely?